<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Basics (Good)</span></div>

## 1 More Python Basics

### 1.1 Computers read = from Right to Left!

In [7]:

# Your code here
x = 40
y = x + 2
print(x)
print('x')

40
x


In [8]:
y = 40
y = y + 2 #works in programming but not in nornal math
print(y)

42


In [9]:
x = y = 10

### 1.2 Shorter and cleaner

In [12]:

# Your code here
#ADDITION
y = 40
y = y + 2
print(y)
print(y)

42
42


In [13]:
print(y)

42


In [14]:
y = 40
y += 2    # Same as y = y + 2
y

42

In [16]:
y #output follows the latest function, not in chronological order

38

In [15]:
#SUBTRACTION
y = 40
y -= 2     # Same as y = y - 2
y

38

In [17]:
#DIVISION
y = 40
y /= 2    # Same as y = y/2
y

20.0

In [18]:
#MULTIPLICATION
y = 40
y *= 2    # Same as y = y * 2
y


80

### 1.3 Asking questions

In [ ]:

# Your code here


In [19]:
#1: CREATE LIST
#2: ASK QNS ABOUT THE LIST

In [20]:
fruits = ["apple", "banana", "pineapple", "jackfruit"]
vegetables = ["celery", "potato", "broccoli", "kale"]

In [21]:
"apple" in fruits


True

In [22]:
"peach" in fruits

False

In [23]:
"peach" not in fruits

True

In [24]:
("apple" in fruits) and ("celery" in vegetables)

True

In [25]:
("apple" in fruits) or ("celery" in vegetables)

True

In [26]:
('apple' in fruits) or ('guava' in vegetables)

True

|Questions/Condition|Math Symbol|Pyton Symbol
|:---|---|---:|
|Equals?|=|==|
|Not equals?|≠|!=|
|Less than?|<|<|
|Greater than?|>|>|
|Less than or equal?|≤|<=|
|Greater than or equal?|≤|>=|

In [27]:
3>10.5

False

In [29]:
'apples' > 'oranges' 
#This comparison works because English letters are internally represented as numbers. 
#For example, a is 97 and o is 111.
#ACSII

False

In [30]:
(x > 5) and (x < 15) 

True

In [31]:
5 < x < 15

True

In [ ]:
#The above 2 lines are equal

### 1.4 There is more to if

In [1]:

# Your code here

name = 'Batman'

if name == 'Batman':
    print('Hello Hero | Batman!')
elif name == 'Robin':
    print('Hello Sidekick | Robin!')
else:
    print('Hello World!')

Hello Hero | Batman!


In [2]:
module='SP2273'
if module=='SP2273':
    print('SP2273-coding module')
elif module =='SP2271':
    print('SP2271-writing module')
else:
    print('Not a module')

SP2273-coding module


In [10]:
module='SP2271'
if module=='SP2273':
    print('SP2273-coding module')
elif module =='SP2271':
    print('SP2271-writing module')
else:
    print('Not a module')

SP2271-writing module


In [11]:
module='SP2274'

if module=='SP2273':
    print('SP2273-coding module')
elif module =='SP2271':
    print('SP2271-writing module')
else:
    print('Not a module')

Not a module


### 1.5 Python stores information in different formats or types.

In [12]:

# Your code here
x = int(1.234)
print(x, type(x))

1 <class 'int'>


In [13]:
x = str(1.234)
print(x, type(x))

1.234 <class 'str'>


In [14]:
x = float(1.234)
print(x, type(x))

1.234 <class 'float'>


In [15]:
x = complex(1.234)
print(x, type(x))

(1.234+0j) <class 'complex'>


In [16]:
#Typecasting-forcing python to change code type

In [17]:
x = '1.234'
print(x, type(x))

1.234 <class 'str'>


In [18]:
x = float(x)
print(x, type(x))

1.234 <class 'float'>


### 1.6 Combining English and variables

In [19]:

# Your code here
name = "Batman"
print(f"Hello {name}!")

Hello Batman!


In [20]:
name = "Batman"
print(f"Hello {name.upper()}!")

Hello BATMAN!


In [24]:
x = 10
print(f"The value of {x} squared is {x**2}!")

The value of 10 squared is 100!


In [25]:
x = 10
print(f"The value of {x} squared is [x**2]!")

The value of 10 squared is [x**2]!


In [22]:
#Note the f and the { } in the above command. This is called f-string or string interpolation.
#Note that [] brackets wont work

In [26]:
text = 'Bruce Wayne is Batman.'
print(f'{text}')

Bruce Wayne is Batman.


In [27]:
print(f'{text:>30}')      # A block of 30 characters; aligned right

        Bruce Wayne is Batman.


In [28]:
print(f'{text:^30}')      # A block of 30 characters; aligned centre

    Bruce Wayne is Batman.    


In [29]:
print(f'{text:<30}')      # A block of 30 characters; aligned left

Bruce Wayne is Batman.        


In [1]:
!conda install numpy.

Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: failed with initial frozen solve. Retrying with flexible solve.

PackagesNotFoundError: The following packages are not available from current channels:

  - numpy.

Current channels:

  - https://repo.anaconda.com/pkgs/main/osx-arm64
  - https://repo.anaconda.com/pkgs/main/noarch
  - https://repo.anaconda.com/pkgs/r/osx-arm64
  - https://repo.anaconda.com/pkgs/r/noarch

To search for alternate channels that may provide the conda package you're
looking for, navigate to

    https://anaconda.org

and use the search bar at the top of the page.




In [2]:
import numpy as np

In [3]:
print(f'The cube of pi to 6 decimal places is {np.pi**3:.6f}')

The cube of pi to 6 decimal places is 31.006277


In [4]:
print(np.pi)

3.141592653589793


In [40]:
x = 'test'

In [42]:
print(f'The cube of pi to 6 decimal places is {x:07.3}')
#The f in .6f is used to tell the f-string to output the number in decimal notation.

The cube of pi to 6 decimal places is tes0000


In [12]:
print(f'The cube of pi to 6 decimal places is {np.pi**3:.6e}') 
#The e in .6e is used to tell the f-string to output the number in scientific notation.

The cube of pi to 6 decimal places is 3.100628e+01


The f-string formatting has the structure {X:>0Y.ZW}. 
So here is what you can use for each of the letters (X,Y,>,0,Z,W) .

W	Specifies the type of variable:
f - float 
d- integer 
s- string 
g - Ask python figure out

X	Variable to format (Can be a number or a string)

>	Alignment Use <, >, ^ for Left, Right and Centre

Y	Total number of characters ALLOCATED for the answer
    if got less than expected, then will fill up with 0's
    For decimals, 3 would not be possible since python forcing to print more than 3 characters hence will override.
    Eg. {3.14159265:3.5} will print 3.14159 instead of 3.14 beacuse python default number of characters is more than the thing we wanted to print.

Z	Number of decimal places

0	Use 0’s to pad the spaces

### 1.7 Python can be a prima-donna.

In [2]:

# Your code here
print hello world #error is the ()

SyntaxError: Missing parentheses in call to 'print'. Did you mean print(...)? (1012304432.py, line 2)

## 2 Best Practices for Scientific Computing

### 2.1 Some tips

In [3]:

# Your code here


I have only given those points that apply to you now. I will bring up more as we progress on our journey.

1. Write programs for people, not computers.
1. Optimise software only after it works correctly.
1. Document design and purpose, not mechanics.
1. Collaborate.

### 2.2 Getting help.

In [ ]:

# Your code here


The internet (e.g. [stack overflow)](https://stackoverflow.com/questions/tagged/python) is the fastest way to get a question about programming answered. However, you can also get information from within Python by:

For example:
    can type (help(print)

## Exercise 1 :  Bias for some heroes

In [2]:
Qn: decode
    def hello(name):
    if name == "Batman":
        print("Hello Batman!")
    else:
        print("Hello World!")


# Your solution here




IndentationError: unexpected indent (2875081110.py, line 2)

In [60]:
name = 'Superman'
print(name == ('Batman' or 'Wonder Woman'))

Wonder Woman


In [72]:
name = 'Superman'
print(name == 'Batman' or 'Wonder Woman')

Wonder Woman


In [73]:
def hello(name):
    if name == 'Batman' or name=='Wonder Woman':
        print(f"Hello {name.upper()}!")
    else:
        print(f"Hello {name.lower()}!")
    
hello('Batman')
hello('Wonder Woman')
hello('Aqua Man')

#python runs name=='Batman' first--> gives false, then runs 'Wonder Woman'--> gives Wonder Woman, 
#then if(false or Wonder Woman)--> always give if(Wonder Woman) because python will always choose non-false
#Hence, only when Wonder Woman is there, then and only then it will print correctly, else is lower case
#Solution: if name==''or name==''NOT if name==''or''

Hello BATMAN!
Hello WONDER WOMAN!
Hello aqua man!


In [ ]:
if name == 'Batman'

In [77]:
def hello(name):
    if name == ('Batman'):
        print(f"Hello {name.upper()}!")
    elif name == ('Wonder Woman'):
        print(f"Hello {name.upper()}!")
    else:
        print(f"Hello {name.lower()}!")

hello('Batman')
hello('Wonder Woman')
hello('Aqua Man')
hello('Iron Man')
'Wonder Woman'


Hello BATMAN!
Hello WONDER WOMAN!
Hello aqua man!
Hello iron man!


'Wonder Woman'

In [80]:
def hello(name):
    if name == 'Batman':
        print(f"Hello {name.upper()}!")
    elif name == 'Black Widow':
        print(f"Hello {name.upper()}!")
    else: 
        print(f'Hello {name.lower()}!')


hello('Batman')    
hello('Black Widow')
hello('Aqua Man')

Hello BATMAN!
Hello BLACK WIDOW!
Hello aqua man!


## Exercise 2 :  Bias for some heroes, continues

In [ ]:
batMan = 5
str(batMan)

In [ ]:



# Your solution here




In [84]:
def hello(name):
    if name == ('Batman'):
        print(f"Hello {name.upper()}!")
    elif name == ('Wonder Woman'):
        print(f"Hello {name.upper()}!")
    else:
        print(f"Hello {name.lower()}!")
    
hello(str(batMan)) #batMan here is a variable and not defined


NameError: name 'batMan' is not defined

In [2]:
def hello(name):
    name = name.upper()
    if name == 'BATMAN':
        print(f"Hello {name.upper()}!")
    elif name == 'WONDER WOMAN':
        print(f"Hello {name.upper()}!")
    else:
        print(f"Hello {name.lower()}!")
        
hello('batMan')
hello('wonDer WOMan')
hello('AQUa maN')

Hello BATMAN!
Hello WONDER WOMAN!
Hello aqua man!


## Exercise 3 :  Look for help to replace

In [ ]:



# Your solution here




In [4]:
string='The Joker lives in Gotham City.'

In [9]:
string.replace( 'Joker', 'Batman' )

'The Batman lives in Gotham City.'

In [10]:
x='I want to sleep.'

In [11]:
x.replace('sleep','study')

'I want to study.'